File: templates/ObjectiveC-UIKit/Features/Home/HomeViewController.m
//
//  HomeViewController.m
//  ObjectiveC-UIKit
//

#import "HomeViewController.h"
#import "SettingsViewController.h"
#import "APIClient.h"
#import "Item.h"
#import "UIView+Extensions.h"

static NSString *const kCellReuseIdentifier = @"HomeItemCell";

@interface HomeViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) UIActivityIndicatorView *loadingIndicator;

// Model backing the table.
@property (nonatomic, copy) NSArray<Item *> *items;

@end

@implementation HomeViewController

#pragma mark - Lifecycle

- (instancetype)init {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _items = @[];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    self.title = @"Home";
    self.view.backgroundColor = [UIColor systemBackgroundColor];

    self.navigationItem.rightBarButtonItem =
        [[UIBarButtonItem alloc] initWithImage:[UIImage systemImageNamed:@"gearshape"]
                                         style:UIBarButtonItemStylePlain
                                        target:self
                                        action:@selector(didTapSettings)];

    [self configureTableView];
    [self configureLoadingIndicator];

    [self reloadItemsShowingSpinner:YES];
}

#pragma mark - View Configuration

- (void)configureTableView {
    UITableView *tableView =
        [[UITableView alloc] initWithFrame:CGRectZero
                                     style:UITableViewStylePlain];
    tableView.dataSource = self;
    tableView.delegate = self;
    [tableView registerClass:[UITableViewCell class]
      forCellReuseIdentifier:kCellReuseIdentifier];

    UIRefreshControl *refreshControl = [[UIRefreshControl alloc] init];
    [refreshControl addTarget:self
                       action:@selector(didPullToRefresh:)
             forControlEvents:UIControlEventValueChanged];
    tableView.refreshControl = refreshControl;

    [self.view addSubview:tableView];
    [tableView pinEdgesToView:self.view];

    self.tableView = tableView;
}

- (void)configureLoadingIndicator {
    UIActivityIndicatorView *indicator =
        [[UIActivityIndicatorView alloc]
            initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleLarge];
    indicator.hidesWhenStopped = YES;
    [self.view addSubview:indicator];
    [indicator centerInView:self.view];
    self.loadingIndicator = indicator;
}

#pragma mark - Actions

- (void)didTapSettings {
    SettingsViewController *settings = [[SettingsViewController alloc] init];
    [self.navigationController pushViewController:settings animated:YES];
}

- (void)didPullToRefresh:(UIRefreshControl *)sender {
    [self reloadItemsShowingSpinner:NO];
}

#pragma mark - Data Loading

- (void)reloadItemsShowingSpinner:(BOOL)showSpinner {
    if (showSpinner) {
        [self.loadingIndicator startAnimating];
    }

    __weak typeof(self) weakSelf = self;
    // Completion is delivered on the main queue by APIClient, so it is safe
    // to update UIKit directly here.
    [[APIClient sharedClient] fetchItemsWithCompletion:^(NSArray<Item *> *_Nullable items,
                                                          NSError *_Nullable error) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (strongSelf == nil) {
            return;
        }

        [strongSelf.loadingIndicator stopAnimating];
        [strongSelf.tableView.refreshControl endRefreshing];

        if (error != nil) {
            [strongSelf presentLoadError:error];
            return;
        }

        strongSelf.items = items ?: @[];
        [strongSelf.tableView reloadData];
    }];
}

- (void)presentLoadError:(NSError *)error {
    UIAlertController *alert =
        [UIAlertController alertControllerWithTitle:@"Couldn't Load Items"
                                            message:error.localizedDescription
                                     preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"Retry"
                                              style:UIAlertActionStyleDefault
                                            handler:^(UIAlertAction *_Nonnull action) {
        [self reloadItemsShowingSpinner:YES];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Dismiss"
                                              style:UIAlertActionStyleCancel
                                            handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    return (NSInteger)self.items.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell =
        [tableView dequeueReusableCellWithIdentifier:kCellReuseIdentifier
                                        forIndexPath:indexPath];

    Item *item = self.items[(NSUInteger)indexPath.row];
    UIListContentConfiguration *content = [cell defaultContentConfiguration];
    content.text = item.title;
    content.secondaryText = item.subtitle;
    cell.contentConfiguration = content;

    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView
    didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    // TODO: Push a detail view controller for the selected item.
}

@end

File: templates/Swift-UIKit/Features/Home/HomeViewController.swift
import UIKit
import Observation

/// Programmatic UIKit view controller for the Home feature.
///
/// Observes the `@Observable` view model using `withObservationTracking` and
/// re-renders whenever any accessed observable property changes.
@MainActor
final class HomeViewController: UIViewController {

    // MARK: - Dependencies

    private let viewModel: HomeViewModel

    // MARK: - UI

    private let tableView = UITableView(frame: .zero, style: .insetGrouped)
    private let refreshControl = UIRefreshControl()
    private let activityIndicator = UIActivityIndicatorView(style: .large)

    private static let cellReuseID = "ItemCell"

    // MARK: - Init

    init(viewModel: HomeViewModel) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Home"
        view.backgroundColor = .systemBackground

        setupTableView()
        setupActivityIndicator()
        observeViewModel()

        Task { await viewModel.load() }
    }

    // MARK: - Setup

    private func setupTableView() {
        tableView.dataSource = self
        tableView.delegate = self
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: Self.cellReuseID)
        tableView.refreshControl = refreshControl
        refreshControl.addTarget(self, action: #selector(handleRefresh), for: .valueChanged)

        view.addSubview(tableView)
        tableView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: view.topAnchor),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        ])
    }

    private func setupActivityIndicator() {
        activityIndicator.hidesWhenStopped = true
        view.addSubview(activityIndicator)
        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            activityIndicator.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            activityIndicator.centerYAnchor.constraint(equalTo: view.centerYAnchor),
        ])
    }

    // MARK: - Observation

    /// Registers a one-shot observation. `withObservationTracking` fires the
    /// `onChange` closure once when any property read in `apply` changes, so we
    /// re-arm it after each render to keep observing.
    private func observeViewModel() {
        withObservationTracking {
            // Read every property we render so the tracker subscribes to them.
            _ = viewModel.items
            _ = viewModel.isLoading
            _ = viewModel.errorMessage
        } onChange: { [weak self] in
            // onChange is delivered off the main actor; hop back before UIKit work.
            Task { @MainActor [weak self] in
                self?.render()
                self?.observeViewModel()
            }
        }
        render()
    }

    private func render() {
        if viewModel.isLoading && viewModel.items.isEmpty {
            activityIndicator.startAnimating()
        } else {
            activityIndicator.stopAnimating()
        }

        if !viewModel.isLoading {
            refreshControl.endRefreshing()
        }

        if let message = viewModel.errorMessage, viewModel.items.isEmpty {
            showError(message)
        }

        tableView.reloadData()
    }

    private func showError(_ message: String) {
        // TODO: Replace with a non-blocking inline error view if preferred.
        guard presentedViewController == nil else { return }
        let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Retry", style: .default) { [weak self] _ in
            Task { await self?.viewModel.load() }
        })
        alert.addAction(UIAlertAction(title: "Dismiss", style: .cancel))
        present(alert, animated: true)
    }

    // MARK: - Actions

    @objc private func handleRefresh() {
        Task { await viewModel.load() }
    }
}

// MARK: - UITableViewDataSource

extension HomeViewController: UITableViewDataSource {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        viewModel.items.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: Self.cellReuseID, for: indexPath)
        let item = viewModel.items[indexPath.row]
        var config = cell.defaultContentConfiguration()
        config.text = item.title
        config.secondaryText = item.subtitle
        cell.contentConfiguration = config
        cell.accessoryType = .disclosureIndicator
        return cell
    }
}

// MARK: - UITableViewDelegate

extension HomeViewController: UITableViewDelegate {

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        // TODO: Push a detail screen for viewModel.items[indexPath.row].
    }
}
